uspto/uspto-to-dolma.py

```diff
@@ -0,0 +1,228 @@
+import argparse
+import glob
+import multiprocessing
+import os
+import sys
+from functools import partial
+from itertools import islice
+from typing import Iterable, Iterator
+
+import polars as pl
+from download_preprocess import parse_html
+from polars import col
+
+from licensed_pile.licenses import PermissiveLicenses
+from licensed_pile.logs import configure_logging, get_logger
+from licensed_pile.write import to_dolma
+
+logger = configure_logging("uspto")
+
+if sys.version_info < (3, 9):
+    raise RuntimeError("Python version >= 3.9 required")
+
+
+def batched(iterable, n):
+    it = iter(iterable)
+    while batch := tuple(islice(it, n)):
+        yield batch
+
+
+def process_datasets(
+    data_dir: str = r"./data/uspto/",
+    url: str = r"http://localhost:3000/convert",
+    limit: int = 0,
+    max_concurrency: int = 4,
+) -> Iterable[dict]:
+    """
+    This function `run_dataset` scans a dataset located in a directory, converts each file in the dataset to a desired
+    format using an API endpoint,and returns an iterable of dictionaries containing the converted data.
+
+    Parameters:
+    - `data_dir` (str): The directory where the dataset is located. Default value is "./data/uspto/".
+    - `url` (str): The API endpoint URL for converting the dataset files. Default value is "http://localhost:3000/convert".
+    - `limit` (int): The maximum number of rows to convert. Default value is 0, which means convert all rows from all files in the dataset.
+    - `max_concurrency` (int): The maximum number of concurrent conversions to perform. Default value is 2.
+
+    Returns:
+    - `Iterable[dict]`: An iterable of dictionaries containing the converted data from each file in the dataset.
+
+    Note:
+    - The `data_dir` parameter should be a valid directory path ending with a forward slash '/'.
+    - The `url` parameter should be a valid API endpoint URL.
+    - The `limit` parameter determines how many row to read. Set it to 0 to convert all files.
+    - The `max_concurrency` parameter determines how many parquet files to process concurrently.
+
+    Example usage:
+    ```python
+    for data in run_dataset(data_dir=r"./data/uspto/", url="http://localhost:3000/convert", limit=10, max_concurrency=2):
+        # Process each converted data entry
+        print(data)
+    ```
+    """
+
+    file_names = glob.glob(os.path.join(data_dir + "*.parquet"))
```

Comment:
Nit: You shouldn't need the os.path.join here right? `data_dir` is typed as a string so it shouldn't do anything?

Reply:
Thanks! Should be good to merge!